src/ILCompiler.ReadyToRun/src/Compiler/DependencyAnalysis/ReadyToRunCodegenNodeFactory.cs

```diff
@@ -729,5 +723,56 @@ protected override IMethodNode CreateUnboxingStubNode(MethodDesc method)
         {
             throw new NotImplementedException();
         }
+
+        struct TypeAndMethod
+        {
+            public readonly TypeDesc Type;
+            public readonly MethodDesc Method;
+
+            public TypeAndMethod(TypeDesc type, MethodDesc method)
+            {
+                Type = type;
+                Method = method;
+            }
+
+            public override bool Equals(object obj)
+            {
+                return obj is TypeAndMethod other && Type == other.Type && Method == other.Method;
+            }
+
+            public override int GetHashCode()
+            {
+                return Type.GetHashCode() ^ unchecked(Method.GetHashCode() * 31);
+            }
+        }
+
+        private Dictionary<TypeAndMethod, ISymbolNode> _delegateCtors = new Dictionary<TypeAndMethod, ISymbolNode>();
+
+        public ISymbolNode DelegateCtor(TypeDesc delegateType, MethodDesc targetMethod, ModuleToken methodToken)
+        {
+            ISymbolNode ctorNode;
+            TypeAndMethod ctorKey = new TypeAndMethod(delegateType, targetMethod);
+            if (!_delegateCtors.TryGetValue(ctorKey, out ctorNode))
```

Comment:
At some point it might be nice for these to start following the pattern that other NodeFactories use (the NodeCache struct instead of the repeated `Dictionary.TryGetValue`/`Add` calls).

Reply:
Some of the dictionaries are trivial to convert but the problem is that in many cases we need to carry around the extra token information. The dictionaries are normally keyed off the final typesystem entities because the tokens may not be generally unique (two typerefs from two modules within the same version bubble may point to a Corelib type, for instance) however, once we're about to create the new instance, we need the token. Is there any way to model this pattern using NodeCache?

Reply:
Right. The fact that the token is not used as part of the key and we just use whatever is the first token we saw it with is a problem by itself (https://github.com/dotnet/corert/pull/6130#discussion_r204315155). We can wait until we hit the problem and then fix the style. This comment can be won't fixed for now.

Reply:
OK, thank you. I have added this item to our R2R work item OneNote page.